### PR TITLE
[Topi][Op][PyTorch][Vitas] Fix inconsistent kernel layout conventions for conv2d_transpose

### PR DESCRIPTION
**Pull request by AndrewZhaoLuo**
The old kernel_layout convention was incorrect for conv2d_transpose. They mixed up `I` and `O`. 

E.g. if we had a conv with 5 input channels and 10 output channels with height and width of 3 the weight for `IOHW` would be `[10, 5, 3, 3]` instead of `[5, 10, 3, 3]`. Likewise for `OIHW` we would have `[5, 10, 3, 3]` instead of `[10, 5, 3, 3]`. This is extremely confusing and several times within the codebase people note the "convention" for "O" and "I" is flipped.  

This is just wrong I believe. `I` and `O` should be related to the input and output channels of the convolution only.

Why was the confusion here in first case? The actual topi assume IOHW kernel inputs and we layout transform to achieve this. It seems for regular convs we are used to OIHW layouts which is why we might have had the confusion. Indeed in PyTorch, conv2d_transpose weights follow IOHW convention:

```
>>> nn.ConvTranspose2d(10, 30, 5, groups=2).state_dict()['weight'].shape
torch.Size([10, 15, 5, 5])
```

While conv2d follow OIHW:

```
>>> nn.Conv2d(10, 30, 5, groups=2).state_dict()['weight'].shape
torch.Size([30, 5, 5, 5])
```

This has caused me a great deal of confusion. It seems like the current code only works because people mix up the layout convention manually. E.g. the layout in the code is said to be OIHW but the shape of the tensor and how they run the calculation is done in IOHW.

This is my attempt at untying the knot and making it much more consistent.

TODOs: 
Make pyxir remove this line and renable the skipped test for vitas.py: https://github.com/Xilinx/pyxir/blob/master/python/pyxir/frontend/tvm/relay_tools/relay_l2_convolution.py#L380

### CODE REVIEW

File: tests/python/contrib/test_vitis_ai/test_vitis_ai_codegen.py [hunk 1]
```diff
@@ -241,6 +240,13 @@ def test_upsampling(dpu_target):
     verify_codegen(mod, dpu_target=dpu_target)
 
 
+@pytest.mark.skip(
+    reason="I and O used to be mixed up in kernel layouts in TVM."
+    "This is fixed, but vitis needs to adopt the new convention."
+    "To change, simply remove this line:"
+    "https://github.com/Xilinx/pyxir/blob/bef661d6d77adcdbd2cf4163f2cf3a1d31d40406/"
+    "python/pyxir/frontend/tvm/relay_tools/relay_l2_convolution.py#L380"
+)
```
Comment (comaniac) [inline tests/python/contrib/test_vitis_ai/test_vitis_ai_codegen.py:249]: cc @jtuyls

**Review by comaniac — APPROVED** [review]
The clarification makes sense to me so I approved first.
It would be better if @masahi could help take a final check.

**Comment by masahi** [comment]
Is the PR https://github.com/apache/tvm/pull/9465 related?

**Comment by junrushao1994** [comment]
@Lyken17 is this related to your fix? Would be great if you could take a look :-) Thanks a lot!

File: tests/python/contrib/test_vitis_ai/test_vitis_ai_codegen.py (hunk 1 above)
Comment (jtuyls) [reply tests/python/contrib/test_vitis_ai/test_vitis_ai_codegen.py:249]: Thanks, this kernel layout switch makes a lot of sense to me. I will fix this in the Vitis flow.

**Comment by Lyken17** [comment]
@junrushao1994 Yes, definitely related. We both noticed that the `conv2d_transpose` weight layout should be `IOHW` rather than `OIHW` (used in `conv2d`). This is indeed confusing and leads to several wrong shape checks. Will go through the code in details later today.

File: src/relay/op/nn/convolution.h [hunk 2]
```diff
@@ -1044,7 +1044,7 @@ bool Conv2DTransposeRel(const Array<Type>& types, int num_inputs, const Attrs& a
   if (data == nullptr) return false;
 
   static const Layout kNCHW("NCHW");
-  static const Layout kOIHW("OIHW");
+  static const Layout kIOHW("IOHW");
```
Comment (Lyken17) [inline src/relay/op/nn/convolution.h:1047]: Does the layout variable affect later calculation, or is it just an alias of layout?

**Review by Lyken17 — COMMENTED** [review]
LGTM. This PR fixs the wrongly adapted `OIHW` in current `topi.nn.conv2d_transpose` as well as many related shape claims / defnitions.

**Comment by junrushao1994** [comment]
Thank you all for the huge effort this week!

**Pull request merged**